pkg/discoveryservice/xdss/stats/collector.go

```diff
@@ -0,0 +1,87 @@
+package stats
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// ensure Stats implements the prometheus Collector interface
+var _ prometheus.Collector = Stats{}
+
+// Descriptors used to create the metrics
+var (
+	requestCountDesc = prometheus.NewDesc(
+		"marin3r_xdss_discovery_requests_total",
+		"Number of discovery requests",
+		[]string{"node_id", "resource_type", "pod_name"}, nil,
+	)
+	ackCountDesc = prometheus.NewDesc(
+		"marin3r_xdss_discovery_ack_total",
+		"Number of discovery ACK responses",
+		[]string{"node_id", "resource_type", "pod_name"}, nil,
+	)
+	nackCountDesc = prometheus.NewDesc(
+		"marin3r_xdss_discovery_nack_total",
+		"Number of discovery NACK responses",
+		[]string{"node_id", "resource_type", "pod_name"}, nil,
+	)
+	infoDesc = prometheus.NewDesc(
+		"marin3r_xdss_discovery_info",
+		"Number of discovery NACK responses",
+		[]string{"node_id", "resource_type", "pod_name", "version"}, nil,
+	)
+)
+
+// Describe is implemented with DescribeByCollect. That's possible because the
+// Collect method will always return the same 4 metrics with the same 4
+// descriptors.
+func (xmc Stats) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(xmc, ch)
+}
+
+// Collect dumps all the keys in the stats cache. Then it
+// creates constant metrics for each modeID/resourceType/pod on the fly based on the
+// dumped returned data.
+func (s Stats) Collect(ch chan<- prometheus.Metric) {
+
+	items := s.DumpAll()
+	for k, v := range items {
+		key := NewKeyFromString(k)
+
+		switch metric := key.StatName + "/" + key.Version; metric {
+
+		case "request_counter/*":
+			ch <- prometheus.MustNewConstMetric(
+				requestCountDesc,
+				prometheus.CounterValue,
+				float64(v.Object.(int64)),
+				key.NodeID, key.ResourceType, key.PodID,
+			)
+
+		case "ack_counter/*":
+			ch <- prometheus.MustNewConstMetric(
+				ackCountDesc,
+				prometheus.CounterValue,
+				float64(v.Object.(int64)),
+				key.NodeID, key.ResourceType, key.PodID,
+			)
+
+		case "ack_counter/" + key.Version:
```

Comment:
inforDesk and ack cases are the same?

Reply:
Nope, they are not the same. We store 2 different ack statistics in the internal cache (the cache is just an in-memory key-value store):

* `<node-id>:<resource-type>:<resource-version>:<pod>:ack_counter`
* `<node-id>:<resource-type>:*:<pod>:ack_counter`

The second one is an aggregation of the first one (like doing a `sum by` with promQL), because we don't want a prometheus metric created for each version of each resource type, that would generate an explosion of metrics in the prometheus database.

The first one though, has all the information about the version the envoy resources of a pod are at, so it's handy to generate a prometheus metric of the info type that would look like:

```
marin3r_xdss_discovery_info{pod=..., node_id=..., resource_type=..., version=...} = 0

```

Using this info metric you can detect drifts between pods of the same node-id being at different versions.

Long story short, the cases are not the same :sweat_smile:

Reply:
haha fair enough